File: tools/event_prep/write_events_in.py
#!/usr/bin/env python3
"""events.json -> events.in, with the irrigation method code normalised.

PEcAn.SIPNET::write.events.SIPNET() maps irrigation methods as
`soil -> 1, everything else -> 0`, so `flood` is written as canopy (0).
SIPNET's own vocabulary is canopy=0, soil=1, flood=2 (see
workflows/tools/write_sipnet_event_file.R). At a flooded rice paddy that
difference is not cosmetic, so the method code is rewritten here from the
JSON, which is the source of truth and schema-valid either way.

Remove this correction once the upstream mapping is fixed; the verification
below will simply find nothing to change.
"""
import json, subprocess, sys, datetime, argparse, re, os

# SIPNET's documented vocabulary is canopy=0, soil=1, flood=2, but the build in
# use here rejects 2 outright ("Unknown irrigation method type: 2") and aborts the
# run. Flood irrigation is therefore emitted as canopy until a binary that
# implements it is available; the JSON keeps method="flood" so the intent is not
# lost, and the anaerobic/waterDrainFrac parameters that would make a flooded
# paddy meaningful are not set in this configuration either.
METHOD_CODE = {"canopy": 0, "soil": 1, "flood": 0}
SIPNET_SUPPORTS_FLOOD = False


def doy(datestr):
    d = datetime.date.fromisoformat(datestr)
    return d.timetuple().tm_yday


def main(events_json, outdir):
    os.makedirs(outdir, exist_ok=True)
    r = subprocess.run(
        ["Rscript", "-e",
         f'suppressMessages(library(PEcAn.SIPNET)); '
         f'PEcAn.SIPNET::write.events.SIPNET("{events_json}","{outdir}")'],
        capture_output=True, text=True)
    if r.returncode != 0:
        sys.exit(f"write.events.SIPNET failed:\n{r.stderr}")

    doc = json.load(open(events_json))
    infile = os.path.join(outdir, f"events-{doc['site_id']}.in")
    lines = open(infile).read().splitlines()

    # (year, doy) -> intended method, from the JSON
    want = {}
    for e in doc["events"]:
        if e["event_type"] == "irrigation":
            y = int(e["date"][:4])
            want[(y, doy(e["date"]))] = METHOD_CODE[e.get("method", "canopy")]

    fixed, checked = 0, 0
    out = []
    for ln in lines:
        m = re.match(r"^(\d+)\s+(\d+)\s+irrig\s+(\S+)\s+(\S+)\s*$", ln)
        if not m:
            out.append(ln); continue
        y, d, amt, code = int(m.group(1)), int(m.group(2)), m.group(3), int(m.group(4))
        checked += 1
        target = want.get((y, d))
        if target is not None and target != code:
            out.append(f"{y}  {d}  irrig  {amt} {target}")
            fixed += 1
        else:
            out.append(ln)
    open(infile, "w").write("\n".join(out) + "\n")

    print(f"  wrote {infile}")
    print(f"  lines {len(out)} | json events {len(doc['events'])}"
          f" {'OK' if len(out) == len(doc['events']) else '<-- MISMATCH'}")
    print(f"  irrigation lines checked {checked}, method code corrected on {fixed}")
    # every irrigation line must now agree with the JSON
    bad = []
    for ln in out:
        m = re.match(r"^(\d+)\s+(\d+)\s+irrig\s+\S+\s+(\d+)\s*$", ln)
        if m:
            k = (int(m.group(1)), int(m.group(2)))
            if want.get(k) is not None and want[k] != int(m.group(3)):
                bad.append(ln)
    print(f"  verification: {'all irrigation methods match the JSON' if not bad else f'{len(bad)} STILL WRONG'}")
    return 0 if not bad else 1


if __name__ == "__main__":
    ap = argparse.ArgumentParser()
    ap.add_argument("events_json")
    ap.add_argument("outdir")
    a = ap.parse_args()
    sys.exit(main(a.events_json, a.outdir))

File: tools/event_prep/write_events_in.R
#!/usr/bin/env Rscript
# events.json -> events.in, rewriting the irrigation method code from the JSON, which
# PEcAn.SIPNET::write.events.SIPNET would otherwise collapse to canopy.

suppressMessages({
  library(jsonlite)
  library(PEcAn.SIPNET)
})

# SIPNET documents canopy=0, soil=1, flood=2, but the build in use rejects 2
# ("Unknown irrigation method type: 2") and aborts, so flood is emitted as canopy.
METHOD_CODE <- c(canopy = 0L, soil = 1L, flood = 0L)

main <- function(events_json, outdir) {
  dir.create(outdir, showWarnings = FALSE, recursive = TRUE)
  PEcAn.SIPNET::write.events.SIPNET(events_json, outdir)

  doc <- jsonlite::fromJSON(events_json, simplifyDataFrame = FALSE)
  infile <- file.path(outdir, paste0("events-", doc$site_id, ".in"))
  lines <- readLines(infile)

  # (year, doy) -> intended method code, from the JSON
  want <- new.env(parent = emptyenv())
  for (e in doc$events) {
    if (!identical(e$event_type, "irrigation")) next
    d <- as.Date(e$date)
    key <- paste(format(d, "%Y"), as.integer(format(d, "%j")))
    m <- if (is.null(e$method)) "canopy" else e$method
    assign(key, unname(METHOD_CODE[[m]]), envir = want)
  }

  pat <- "^(\\d+)\\s+(\\d+)\\s+irrig\\s+(\\S+)\\s+(\\S+)\\s*$"
  checked <- 0L; fixed <- 0L
  out <- vapply(lines, function(ln) {
    if (!grepl(pat, ln)) return(ln)
    g <- regmatches(ln, regexec(pat, ln))[[1]]
    yr <- g[2]; dy <- as.integer(g[3]); amt <- g[4]; code <- as.integer(g[5])
    checked <<- checked + 1L
    key <- paste(yr, dy)
    if (!exists(key, envir = want, inherits = FALSE)) return(ln)
    target <- get(key, envir = want, inherits = FALSE)
    if (identical(target, code)) return(ln)
    fixed <<- fixed + 1L
    sprintf("%s  %d  irrig  %s %d", yr, dy, amt, target)
  }, character(1), USE.NAMES = FALSE)

  writeLines(out, infile)

  cat("  wrote", infile, "\n")
  cat("  lines", length(out), "| json events", length(doc$events),
      if (length(out) == length(doc$events)) "OK" else "<-- MISMATCH", "\n")
  cat("  irrigation lines checked", checked, ", method code corrected on", fixed, "\n")

  bad <- 0L
  for (ln in out) {
    if (!grepl(pat, ln)) next
    g <- regmatches(ln, regexec(pat, ln))[[1]]
    key <- paste(g[2], as.integer(g[3]))
    if (exists(key, envir = want, inherits = FALSE) &&
        !identical(get(key, envir = want, inherits = FALSE), as.integer(g[5]))) bad <- bad + 1L
  }
  cat("  verification:", if (bad == 0L) "all irrigation methods match the JSON"
      else paste(bad, "STILL WRONG"), "\n")
  invisible(if (bad == 0L) 0L else 1L)
}

args <- commandArgs(trailingOnly = TRUE)
if (length(args) != 2L) stop("usage: write_events_in.R <events.json> <outdir>")
quit(status = main(args[1], args[2]))
